src/storage/src/types/errors.rs

```diff
@@ -50,22 +50,200 @@ impl Codec for DecodeError {
         "protobuf[DecodeError]".into()
     }
 
-    fn encode<B: BufMut>(&self, buf: &mut B) {
+    fn encode<B>(&self, buf: &mut B)
+    where
+        B: BufMut,
+    {
         self.into_proto()
             .encode(buf)
-            .expect("no required fields means no initialization errors");
+            .expect("no required fields means no initialization errors")
     }
 
-    fn decode(buf: &[u8]) -> Result<Self, String> {
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, String> {
         let proto = ProtoDecodeError::decode(buf).map_err(|err| err.to_string())?;
         proto.into_rust().map_err(|err| err.to_string())
     }
 }
 
 impl Display for DecodeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match &self.raw {
+            None => write!(f, "{}", self.inner),
+            Some(raw) => write!(f, "{} (original bytes: {:x?})", self.inner, raw),
+        }
+    }
+}
+
+#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash)]
+pub enum DecodeErrorInner {
```

Comment:
I think `Inner` is not a great name for a public api, I suggest something like `DecodeError`+`DecodeErrorKind`

Reply:
same change on the proto, `inner` -> `kind`